pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionMetaDataManager.java

```diff
@@ -174,89 +188,127 @@ public synchronized boolean containsFunction(String tenant, String namespace, St
+            log.error("Could not write into Function Metadata topic", e);
+            errorNotifier.triggerError(e);
+            throw new IllegalStateException("Internal Error updating function at the leader", e);
+        }
+        if (needsScheduling) {
+            this.schedulerManager.schedule();
         }
-
-        FunctionMetaData newFunctionMetaData = FunctionMetaDataUtils.generateUpdatedMetadata(existingFunctionMetadata, functionMetaData);
-
-        Request.ServiceRequest updateRequest = ServiceRequestUtils.getUpdateRequest(
-                this.workerConfig.getWorkerId(), newFunctionMetaData);
-
-        return submit(updateRequest);
     }
 
-
     /**
-     * Sends a deregister request to the FMT (Function Metadata Topic) for a function
-     * @param tenant the tenant the function that needs to be deregistered belongs to
-     * @param namespace the namespace the function that needs to be deregistered belongs to
-     * @param functionName the name of the function
-     * @return a completable future of when the deregister has been applied
+     * Called by the leader service when this worker becomes the leader.
+     * We first get exclusive producer on the metadata topic. Next we drain the tailer
+     * to ensure that we have caught up to metadata topic. After which we close the tailer.
+     * Note that this method cannot be syncrhonized because the tailer might still be processing messages
      */
-    public synchronized CompletableFuture<RequestResult> deregisterFunction(String tenant, String namespace, String functionName) {
-        FunctionMetaData functionMetaData = this.functionMetaDataMap.get(tenant).get(namespace).get(functionName);
-
-        FunctionMetaData newFunctionMetaData = FunctionMetaDataUtils.generateUpdatedMetadata(functionMetaData, functionMetaData);
-
-        Request.ServiceRequest deregisterRequest = ServiceRequestUtils.getDeregisterRequest(
-                this.workerConfig.getWorkerId(), newFunctionMetaData);
+    public void acquireLeadership() {
+        log.info("FunctionMetaDataManager becoming leader by creating exclusive producer");
+        FunctionMetaDataTopicTailer tailer = internalAcquireLeadership();
+        // Now that we have created the exclusive producer, wait for reader to get over
+        if (tailer != null) {
+            try {
+                tailer.stopWhenNoMoreMessages().get();
+            } catch (Exception e) {
+                log.error("Error while waiting for metadata tailer thread to finish", e);
+                errorNotifier.triggerError(e);
+            }
+            tailer.close();
+        }
+        log.info("FunctionMetaDataManager done becoming leader");
+    }
 
-        return submit(deregisterRequest);
+    private synchronized FunctionMetaDataTopicTailer internalAcquireLeadership() {
+        if (exclusiveLeaderProducer == null) {
+            try {
+                exclusiveLeaderProducer = pulsarClient.newProducer()
```

Comment:
couldn't there be a race condition between here and "start()" where you are relying on whether exclusiveLeaderProducer is null or not.  "acquireLeadership()" is called by the "becomeActive" which is using the client listener thread while "start()" is executed by the worker "main" thread

Reply:
nvm